klippa/src/post.rs

```diff
@@ -0,0 +1,128 @@
+//! impl subset() for post
+
+use std::collections::HashMap;
+
+use crate::{Plan, SubsetError, SubsetError::SubsetTableError, SubsetFlags};
+use write_fonts::{
+    read::{
+        tables::post::{Post, DEFAULT_GLYPH_NAMES},
+        FontRef, TableProvider, TopLevelTable,
+    },
+    types::GlyphId,
+    FontBuilder,
+};
+
+// reference: subset() for post in harfbuzz
+// https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-post-table.hh#L96
+pub(crate) fn subset_post(
+    font: &FontRef,
+    plan: &Plan,
+    builder: &mut FontBuilder,
+) -> Result<(), SubsetError> {
+    let post = font.post().or(Err(SubsetTableError(Post::TAG)))?;
+    let mut out = Vec::with_capacity(post.offset_data().len());
+    // copy header
+    out.extend_from_slice(post.offset_data().as_bytes().get(0..32).unwrap());
+
+    let glyph_names = plan
+        .subset_flags
+        .contains(SubsetFlags::SUBSET_FLAGS_GLYPH_NAMES);
+    //version 3 does not have any glyph names
+    if !glyph_names {
+        let major_version = 0x3_u16.to_be_bytes();
+        out.get_mut(0..2).unwrap().copy_from_slice(&major_version);
+        out.get_mut(2..4).unwrap().fill(0);
+    }
+
+    if glyph_names && post.version().to_major_minor() == (2, 0) {
+        subset_post_v2tail(&post, plan, &mut out);
+    }
+
+    builder.add_raw(Post::TAG, out);
+    Ok(())
+}
+
+fn subset_post_v2tail(post: &Post, plan: &Plan, out: &mut Vec<u8>) {
+    //copy numGlyphs
+    out.extend_from_slice(&(plan.num_output_glyphs as u16).to_be_bytes());
+
+    //init all glyphNameIndex as 0
+    let idx_start = out.len();
+    let new_len = out.len() + plan.num_output_glyphs * 2;
+    out.resize(new_len, 0);
+
+    let max_old_gid = plan.glyphset.last().unwrap().to_u32() as usize;
+    // for standard glyphs: name indices < 258
+    let glyph_index_iter = post
+        .glyph_name_index()
+        .unwrap()
```

Comment:
this will crash on a malformed post table, is that okay?

Reply:
good catch, fixed it